csharp/src/Google.Protobuf/WritingPrimitives.cs

```diff
@@ -187,29 +189,23 @@ public static void WriteString(ref Span<byte> buffer, ref WriterInternalState st
 
                         // Process 4 chars at a time until there are less than 4 remaining.
                         // We already know all characters are ASCII so there is no need to validate the source.
-                        int currentIndex = 0;
-                        int lastIndexWhereCanReadFourChars = (2 * value.Length - 8);
+                        int lastIndexWhereCanReadFourChars = value.Length - 4;
```

Comment:
This is not the "optimal" length isn't it?
Assuming you have 39 characters, then you only do the optimized code for 32 characters because in the next loop 32+4 is 36 which is more than 35 (39-4), even though 36 chars could be used for your optimized code.
I think something like value.Length - (value.Length & 3) should be more correct.

Reply:
It is a do/while loop. When the numbers of chars processed is 36, then it will stop because there aren't 4 chars remaining.

Reply:
Okay, and what happens with 40 chars?

Reply:
All chars are processed in the do/while loop

Reply:
Oh, forget it, you already wrote it before, you do the do/while loop. Yeah, you're right, sorry for the ruckus.